homeassistant/components/unifiprotect/binary_sensor.py

```diff
@@ -0,0 +1,271 @@
+                "Adding binary sensor entity %s",
+                (description.name or "{index}").format(index=index),
+            )
+
+    return entities
+
+
+class ProtectDeviceBinarySensor(ProtectDeviceEntity, BinarySensorEntity):
+    """A UniFi Protect Device Binary Sensor."""
+
+    def __init__(
+        self,
+        data: ProtectData,
+        description: ProtectBinaryEntityDescription,
+        device: Camera | Light | Sensor | None = None,
+    ) -> None:
+        """Initialize the Binary Sensor."""
+
+        if device and not hasattr(self, "device"):
+            self.device: Camera | Light | Sensor = device
+        self.entity_description: ProtectBinaryEntityDescription = description
+        super().__init__(data)
+        self._doorbell_callback: TaskClass | None = None
+
+    @callback
+    def _async_update_extra_attrs_from_protect(self) -> dict[str, Any]:
+        attrs: dict[str, Any] = {}
+        key = self.entity_description.key
+
+        if key == _KEY_DARK:
+            return attrs
+
+        if key == _KEY_DOORBELL:
+            assert isinstance(self.device, Camera)
+            attrs[ATTR_LAST_TRIP_TIME] = self.device.last_ring
+        elif isinstance(self.device, Sensor):
+            if key in (_KEY_MOTION, _KEY_DOOR):
+                if key == _KEY_MOTION:
+                    last_trip = self.device.motion_detected_at
+                else:
+                    last_trip = self.device.open_status_changed_at
+
+                attrs[ATTR_LAST_TRIP_TIME] = last_trip
+        elif isinstance(self.device, Light):
+            if key == _KEY_MOTION:
+                attrs[ATTR_LAST_TRIP_TIME] = self.device.last_motion
+
+        return attrs
+
+    @callback
+    def _async_update_device_from_protect(self) -> None:
+        super()._async_update_device_from_protect()
+
+        assert self.entity_description.ufp_value is not None
+
+        self._attr_extra_state_attributes = (
+            self._async_update_extra_attrs_from_protect()
+        )
+
+        if self.entity_description.key == _KEY_DOORBELL:
+            last_ring = get_nested_attr(self.device, self.entity_description.ufp_value)
+            now = utcnow()
+
+            is_ringing = (
+                False if last_ring is None else (now - last_ring) < RING_INTERVAL
+            )
+            if is_ringing:
+                if self._doorbell_callback is not None:
+                    self._doorbell_callback.cancel()
+                self._doorbell_callback = asyncio.ensure_future(
+                    self._async_wait_for_doorbell(last_ring + RING_INTERVAL)
+                )
+            self._attr_is_on = is_ringing
+        else:
+            self._attr_is_on = get_nested_attr(
+                self.device, self.entity_description.ufp_value
+            )
+
+    @callback
+    async def _async_wait_for_doorbell(self, end_time: datetime) -> None:
+        _LOGGER.debug("Doorbell callback started")
+        while utcnow() < end_time:
```

Comment:
Can we use async_call_later here instead and cancel and previous timers?

Reply:
Maybe async track time utc would be better.

Reply:
I think `async_call_later` would work better. Then I can use the `async_fire_time_changed` for the tests, right?

Reply:
Also that work with fire time changed

Reply:
Be sure to store the collable that it returns and cancel it if another event happens

Reply:
Figured out my issue with `async_fire_time_changed`, it does not change the system clock if you do that. Haha.